spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java

```diff
@@ -305,40 +327,42 @@ public long count(Class entityClass) {
 
 	@Override
 	public <T> T performReadWriteTransaction(Function<SpannerTemplate, T> operations) {
-		return this.databaseClient.readWriteTransaction()
-				.run(new TransactionCallable<T>() {
-					@Nullable
-					@Override
-					public T run(TransactionContext transaction) { // @formatter:off
-						ReadWriteTransactionSpannerTemplate transactionSpannerTemplate =
-										new ReadWriteTransactionSpannerTemplate(
-										// @formatter:on
+		return this.databaseClient
+				.readWriteTransaction()
+				.run(
+						new TransactionCallable<T>() {
+							@Nullable
+							@Override
+							public T run(TransactionContext transaction) {
+								ReadWriteTransactionSpannerTemplate transactionSpannerTemplate
+										= new ReadWriteTransactionSpannerTemplate(
 										SpannerTemplate.this.databaseClient,
 										SpannerTemplate.this.mappingContext,
 										SpannerTemplate.this.spannerEntityProcessor,
 										SpannerTemplate.this.mutationFactory,
 										SpannerTemplate.this.spannerSchemaUtils,
 										transaction);
-						return operations.apply(transactionSpannerTemplate);
-					}
-				});
+								return operations.apply(transactionSpannerTemplate);
+							}
+						});
 	}
 
 	@Override
-	public <T> T performReadOnlyTransaction(Function<SpannerTemplate, T> operations,
-			SpannerReadOptions readOptions) {
-		SpannerReadOptions options = readOptions == null ? new SpannerReadOptions()
-				: readOptions;
+	public <T> T performReadOnlyTransaction(
+			Function<SpannerTemplate, T> operations, SpannerReadOptions readOptions) {
+		SpannerReadOptions options = readOptions == null ? new SpannerReadOptions() : readOptions;
 		try (ReadOnlyTransaction readOnlyTransaction = options.hasTimestamp()
 				? this.databaseClient.readOnlyTransaction(
 						TimestampBound.ofReadTimestamp(options.getTimestamp()))
 				: this.databaseClient.readOnlyTransaction()) {
-			return operations.apply(new ReadOnlyTransactionSpannerTemplate(
-					SpannerTemplate.this.databaseClient,
-					SpannerTemplate.this.mappingContext,
-					SpannerTemplate.this.spannerEntityProcessor,
-					SpannerTemplate.this.mutationFactory,
-					SpannerTemplate.this.spannerSchemaUtils, readOnlyTransaction));
+			return operations.apply(
```

Comment:
I'm totally new to Spring AOP and transactions so please forgive if I say something that doesn't make sense:

If somebody were to use `performReadOnlyTransaction` or `performReadWriteTransaction` from inside another method annotated with `@Transactional`, then I think that corresponds to the "NESTED" propagation setting : https://docs.spring.io/spring/docs/3.1.x/javadoc-api/org/springframework/transaction/annotation/Propagation.html#NESTED

However, currently in the PR we are getting a TransactionContext that is independent of the ongoing transaction.

Reply:
My recommendation: I think in this function we should be checking if we are already inside an ongoing `@Transactional` transaction and throwing an exception saying Spanner doesn't support transactions-within-transactions.

The 2 alternatives I don't think make sense:

1. current situation where it's an independent transaction. If the containing method got rolled back, then transaction run via the anonymous function inside wouldn't be rolled back, and that is counter-intuitive. 

2. merging the anonymous function into the ongoing transaction. I think this is also counter-intuitive because it would mean the user putting code into an anonymous function is no different than writing it plainly in the containing method's body. it would also make the very suggestive method name `performRead[only/write]Transaction` meaningless, which is unexpected.